tests/components/qwikswitch/test_init.py

```diff
@@ -106,3 +114,232 @@ async def test_sensor_device(hass, aioclient_mock):
     assert state_obj.state == "416"
 
     listen_mock.stop()
+
+
+async def test_switch_device(hass, aioclient_mock, qs_devices):
+    """Test a switch device."""
+
+    async def get_devices_json(method, url, data):
+        return AiohttpClientMockResponse(method=method, url=url, json=qs_devices)
+
+    config = {"qwikswitch": {"switches": ["@a00001"]}}
+    aioclient_mock.get("http://127.0.0.1:2020/&device", side_effect=get_devices_json)
+    listen_mock = MockLongPollSideEffect()
+    aioclient_mock.get("http://127.0.0.1:2020/&listen", side_effect=listen_mock)
+    assert await async_setup_component(hass, QWIKSWITCH, config)
+    hass.bus.async_fire(EVENT_HOMEASSISTANT_START)
+    await hass.async_block_till_done()
+
+    state_obj = hass.states.get("switch.switch_1")
+    assert state_obj.state == "off"
+
+    aioclient_mock.mock_calls.clear()
+    aioclient_mock.get("http://127.0.0.1:2020/@a00001=100")
+    await hass.services.async_call(
+        "switch", "turn_on", {"entity_id": "switch.switch_1"}, blocking=True
+    )
+    await asyncio.sleep(0.01)
+    assert (
+        "GET",
+        URL("http://127.0.0.1:2020/@a00001=100"),
+        None,
+        None,
+    ) in aioclient_mock.mock_calls
+
+    qs_devices[0]["val"] = "ON"
+    state_obj = hass.states.get("switch.switch_1")
+    assert state_obj.state == "off"
+    listen_mock.queue_response(json={"id": "@a00001", "cmd": ""},)
+    await hass.async_block_till_done()
+    state_obj = hass.states.get("switch.switch_1")
+    assert state_obj.state == "on"
+
+    aioclient_mock.mock_calls.clear()
+    aioclient_mock.get("http://127.0.0.1:2020/@a00001=0")
+    await hass.services.async_call(
+        "switch", "turn_off", {"entity_id": "switch.switch_1"}, blocking=True
+    )
+    assert (
+        "GET",
+        URL("http://127.0.0.1:2020/@a00001=0"),
+        None,
+        None,
+    ) in aioclient_mock.mock_calls
+
+    listen_mock.stop()
+
+
+async def test_light_device(hass, aioclient_mock, qs_devices):
+    """Test a light device."""
+
+    async def get_devices_json(method, url, data):
+        return AiohttpClientMockResponse(method=method, url=url, json=qs_devices)
+
+    config = {"qwikswitch": {}}
+    aioclient_mock.get("http://127.0.0.1:2020/&device", side_effect=get_devices_json)
+    listen_mock = MockLongPollSideEffect()
+    aioclient_mock.get("http://127.0.0.1:2020/&listen", side_effect=listen_mock)
+    assert await async_setup_component(hass, QWIKSWITCH, config)
+    hass.bus.async_fire(EVENT_HOMEASSISTANT_START)
+    await hass.async_block_till_done()
+
+    state_obj = hass.states.get("light.dim_3")
+    assert state_obj.state == "on"
+    assert state_obj.attributes["brightness"] == 255
+
+    aioclient_mock.mock_calls.clear()
+    aioclient_mock.get("http://127.0.0.1:2020/@a00003=0")
```

Comment:
Same, needs a `data`

Reply:
same question as above. this doesn't return anything. just setting the mock for when the library calls this url

Reply:
Hopefully explanation above makes sense

Reply:
added some comments to explain what each step is testing. this step just checks that if hass asks the component to turn on a light (or switch), the library sends the right command to the http server. it doesn't need to check that something changed

Reply:
All set commands should return a dict with a `data` key. If you don’t the library creates tasks in the background to retry... 

```python
aioclient_mock.get("http://127.0.0.1:2020/@a00001=on/off", {"id":"@a00001","data":"anything"})
```

Please don’t try to work against the library. If you want to test different scenarios, do it in different test functions.

Reply:
fixed in the latest. however, the library sends @a00001=0. do i need the id as well in the response? it does update the switch

Reply:
i have the data in the response, but not the id. do i need the id as well?

Reply:
The link I shared with you in the beginning only checked that ‘data’ is there and is not ‘NO REPLY’, so no need for the id it seems

https://github.com/kellerza/pyqwikswitch/blob/master/pyqwikswitch/async_.py#L119

Reply:
i know. i saw it too in the library code, but just wondered if you want them both.

Reply:
sounds like data is enough. thanks